### ISSUE TITLE
Proposed Conformance Model

### ISSUE THREAD

**Issue by jeff-zucker** [solid/webid-profile#40]
[EDIT - updated based on Editor's meeting of 2022-09-06, but not including all the comments since last update.  This model is likely to change]

Given that IdPs may wish to restrict how WebID Documents behave, including limiting the ability of the WebID owner to write to the document, we need to support the ESS-like model in which the WebID document is not the same document as the Profile document. 

If we proceed as I've outlined below, we would have two shapes which can occur in one document or in two documents.  In the NSS-like instance where the same document is both the WebID document and the Profile document, it would need to conform to both shapes. In the ESS-like instance the WebID Document could be very minimal and the Solid Profile Document would conform to the same shape as all other Solid Profile Documents.

I also propose a `solid:profileDocument` predicate.  We can add a note that, for backward compatibility, developers may wish to also check for the foaf:primaryTopicOf predicate. 

### The Solid WebID Document

This document should follow the rules set forth in the [WebID Specification](https://www.w3.org/2005/Incubator/webid/spec/identity/) including (TBD:get exact wording)

And it should follow this rule, introduced by this spec

* it MUST either conform to the rules listed below for a Solid Profile Document or else have a triple equivalent to `<?WebID> solid:profileDocument <?SolidProfileDocument>.`. For backward compatibility, the predicate can be foaf:primaryTopicOf instead.

In order to be usable for purposes of verifying identity, a WebID document should also contain predicates mandated by the identity protocol in use.  For example if the [Solid-OIDC ](TBD) protocol is in use, the WebID document MUST include a triple equavent to `<?WebID> solid:oidcIssuer <?Issuer>.`  See the Solid-OIDC specification for details

### The Solid Profile Document

* there MUST be exactly one Solid Profile Document for any given WebID URI
* it MUST have exactly one `space:preferencesFile` triple
* it MUST be writable by the WebID owner or those they designate
* it SHOULD have exactly one `ldp:inbox` triple
* it SHOULD have one or more `space:storage` triples
* it MAY have zero or more triples with the `solid:community` predicate
* it MAY have zero or more triples with the `rdfs:seeAlso` predicate
*  it can be the same document as the WebID Document or it can be a separate Document. In the case in which the Solid Profile Document and the Solid WebID Document are the same, the document MUST conform to the rules listed above for both types of documents.
 
### The Preferences Document (`space:preferencesFile`)

* there MUST be exactly one Preferences Document
* it MUST be writable by the WebID owner or those they designate
* it MUST NOT be readable or writable by anyone other than the WebID owner or those they designate
* it MAY have zero or more triples with the `rdfs:seeAlso` predicate

### Extended Profile Documents (`rdfs:seeAlso`)

* there MAY be zero of more extended profile documents
* they MAY be accessible to public, restricted, or private audiences

Note: Apps can be expected to examine triples in seeAlso documents with the WebID as subject.  They are free to but not required to examine other triples. 

Note: Apps are free to but are not required to examine triples in seeAlso documents using the `rdfs:seeAlso` predicate.

**Comment by jeff-zucker** [comment]
All of which yields this updated diagra
![discovery-new](https://user-images.githubusercontent.com/44732708/184708750-6f9e7e2f-2ef5-4c4e-ad4d-c7a059345d42.png)
m

**Comment by ThisIsMissEm** [comment]
@jeff-zucker I think you've some typos in:
> it MUST either label itself as a Solid Profile Document using a triple equivalent to <> a solid:ProifleDocument. or include a pointer to a Solid Profile Document using a triple equivalent to <?WebID> solid:profileDocument <?SolidProfileDocument>..

Did you mean:
> it MUST either label itself as a Solid Profile Document using a triple equivalent to `<> a solid:ProfileDocument`. or include a pointer to a Solid Profile Document using a triple equivalent to `<?WebID> solid:profileDocument <?SolidProfileDocument>.`.

**Comment by jeff-zucker** [comment]
@ThisIsMissEm - do you mean the missing `tic` marks or something else?

**Comment by gibsonf1** [comment]
Excellent work, this proposal looks really good and solves some key issues around webid standards and discovery.

**Comment by ThisIsMissEm** [comment]
@jeff-zucker would also be good to see how this pairs with pim:storage

**Comment by jeff-zucker** [comment]
@ThisIsMissEm - sorry my eyes are not good enough to see the typo you are talking about.  We mention space:storage (same as pim: but other specs seem to use space:) and say that it SHOULD go in the Solid Profile Document unless there are reasons otherwise.  In other words, if ESS wants to write it in the WebID document instead, it can but if someone needs to write it, they should write it to the Profile Document.  Personally, I think it is a mistake to put this in the WebID document because it forces the public listing of the storage.  Better would be to leave it up to the user - it goes in the Solid Profile by default but they can move it to somewhere private or restricted if they want.

**Comment by bourgeoa** [comment]
For historical reasons, I feel that there should be a 2 step process :
- the old NSS/CSS type could be referenced as 0.9
- new project stating changes and some (if any) migration consideration

This will help developers and server/app maintainers.

**Comment by jeff-zucker** [comment]
I'm afraid I don't understand what you mean by migration.  Both types of
profiles will continue to exist. They are both in existence now, this spec
just takes that into account.  I also do not understand what you mean by 2
step process. What would the steps be?

**Comment by elf-pavlik** [comment]
> Given that IdPs may wish to restrict how WebID Documents behave, including limiting the ability of the WebID owner to write to the document, we need to support the ESS-like model in which the WebID document is not the same document as the Profile document.

Can I find somewhere more details about the *ESS-like* approach and the reasoning behind having a separate WebID document and Profile document?

**Comment by Otto-AA** [comment]
The proposal sounds good to me, thanks for working on this @jeff-zucker. The only concern I have here is the number of requests, as all of them have to be done sequentially which requires multiple round trips between the client and the pod. However, I think this is acceptable, because we only need to do this once and then can cache it for some time.

**Comment by jeff-zucker** [comment]
@elf-pavlik - the reasoning behind the approach, as I understand it is that the Identity provider and the Storage provider can be different entities and that the Identity provider may or may not allow writing to the WebID Document.  Identity providers should have very few restrictions on the WebID document because they may be serving a variety of protocols.  Additionally, some IdPs feel there is a security concern with having the solid:oidcIssuer predicate in a user-writable document.  So, in order to put the minimum requirements on the Identity Provider, we say only that the WebID document  needs to point to a document that is writable by the WebID owner, not that it itself be directly writable.

**Comment by elf-pavlik** [comment]
Thank you @jeff-zucker 

> Additionally, some IdPs feel there is a security concern with having the solid:oidcIssuer predicate in a user-writable document.

I understand that WebID Document has specific security requirements, and possibly not be handled by ACP/WAC rules, and have dedicated IdP provided UI to modify it.
At the same time, I'm not convinced about locking the WebID owner from any modifications to that document. We have discussed that some Resource Servers (storages) may require authentication from a very specific OP (OIDC Issuer), ACP has `acp:issuer` matcher designed for that. 

----

Another question would be on the extensibility of the WebID Document, for example, `solid:oidcIssuer` appears here even though [Solid-OIDC](https://solidproject.org/TR/oidc) is the spec that requires it. 

We have a very similar case with `interop:hasAuthorizationAgent`, required by [Solid Application Interoperability](https://solid.github.io/data-interoperability-panel/specification/)

I don't understand which specification would have the other as a normative dependency. And if `solid:oidcIssuer` would be mandated by this spec, should we also submit `interop:hasAuthorizationAgent` here? We can think of it as AuthZ counterpart for the already included AuthN predicate.

**Comment by jeff-zucker** [comment]
I personally also agree that it may not make a lot of sense to lock the WebID owner out of modifying the WebID document but I don't see how we can be in the position to require servers to allow it.

Good point about the solid:oidcIssuer.  My thought is that the Solid-OIDC spec is the normative one and that we refer to it.  All of the conformance rules for the WebID document with the exception of the solid:solidProfile predicate are based on Solid-OIDC and we should be explicit about that.

**Comment by jeff-zucker** [comment]
> should we also submit `interop:hasAuthorizationAgent` here? We can think of it as AuthZ counterpart for the already included AuthN predicate.

No, I don't believe it should be part of this spec.  Our spec defines a very basic profile.  We will enhance the final section "Other Predicates"  to say that our spec is only one of the specs relating to profiles and that other specs may introduce further predicates and conformance rules required for its own purposes.  We may name specific other specs here, or we may not.

This way our spec, the interoperability spec, the upcoming indexes spec, and future specs on personal profiles and organizational profiles will operate independently of each other.  None of them should contradict each other i.e. forbidding something required by another spec, but other than that, I don't think they should really mention each other.  This leaves developers and users choices.  For example an app based on public data whose main purpose is to share data can work just fine without an authorization agent but, as you mention, that may not be the case for medical or financial apps.  As you say, we should alert developers about security considerations, but in the end, it is up to them what they do about those considerations.

**Comment by jeff-zucker** [comment]
@elf-pavlik  - my view is that this spec defines the expected shape and discovery method for some core profile documents.  It does not address a) what indexes exist b) how apps are authorized c) what individuals and organizations can/should say about themselves.   Other specs will deal with those topics.

**Comment by scenaristeur** [comment]
If a webid doc can 
"include a pointer to a Solid Profile Document", everyone can switch to any Solid profile doc. 
Today my webid doc can point to "a Solid app developer in France"  tomorrow to a little Chinese girl in Mexico" and the day after  my webid can point to "@timBL profile" . How can we know who is behind a webid doc. Can we be sure of any consistency? 
Personnaly I'm really disappointed with the way 'inrupt devs " impose their point of view to the community. I've been struggling with the remplacement of solid-auth-client by the inrupt new auth, there is another discussion about what should developer use wack/ACP and now big changes on the profile doc. I'm not afraid of changes but I think it is not a good way to develop something for the NSS and then push it as a spec for the community. There are big problem that Inrupt devs could help instead of changing what is working. I remember for example a longtime issue where @aveltens post a chat message instead of Tim, the fact that when we autorisé an app, it has acces to all the pod..

**Comment by jeff-zucker** [comment]
> how can we know who is behind a webid doc. Can we be sure of any consistency?

The WebID doc must contain a solid:oidcIssuer predicate that points to an Identity Provider capable of verifying that a logged in user either is or isn't the person registered with them as the owner of that WebID.  That part can be verified.   The WebID document can point to a profile which says it belongs to the Queen of Mars.  That part can't be verified, people can portray their identity any way they want.  That has always been true, this spec does nothing to alter that situation.

I certainly hear your points about the new auth, the differences between ACP/ACL and other aspects Inrupt has introduced - these have all been extremely frustrating for app implementers.  OTOH, I do believe all of those changes have made Solid more secure.   

The change proposed here, in this issue will not mean any change at all for any server or app other than those on ESS and on future servers designed the way ESS is.  It is NOT suggested here to accomodate Inrupt, rather because it is really to everyone's benefit for Identity Providers to face as few constraints as possible.  Identity Providers may be using many different identity protocols, Solid being only one.  They also may have security reasons not to want users to write directly to the WebID document.  So if they can create a WebID document that can't be written to directly, we need a way to say, look over there for the writable stuff.  This lets the Identity Provider control and secure the WebID document but lets the user have full control over their own profile.

**Comment by jeff-zucker** [comment]
@scenaristeur  wrote "everyone can switch to any Solid profile doc"

But there can only be one Solid Profile Document per WebID and it is guaranteed to be either the document at the WebID URI or a document pointed to in that document with the solid:profileDocument predicate.  Apps should not believe something is a Solid Profile Document just because it says it is (and maybe that's a reason to not introduce the SolidProfile class) - they should believe it is a Solid Profile because they got there either directly or one-hop-away from the WebID URI.

**Comment by elf-pavlik** [comment]
>  Apps should not believe something is a Solid Profile Document just because it says it is (and maybe that's a reason to not introduce the SolidProfile class) - they should believe it is a Solid Profile because they got there either directly or one-hop-away from the WebID URI.

One could only honor statements that have the WebID as the subject or object. I think this is a broader issue with the provenance of statements which needs to be addressed when the application is merging multiple graphs/datasets.

> This way our spec, the interoperability spec, the upcoming indexes spec, and future specs on personal profiles and organizational profiles will operate independently of each other. None of them should contradict each other i.e. forbidding something required by another spec, but other than that, I don't think they should really mention each other. 

I think discussion in #43 is relevant, especially what @RubenVerborgh wrote:

> However, we might still want to discuss where that description will be written down. Essentially, the question is whether we want:
>*  the Profile document to have a dependency on the Type Indexes document
>*  the Type Indexes document to have a dependency on the Profile document
>
> I think the second option is the more sustainable one; indexes will still evolve, but the profile should remain constant.

If I understand him correctly I would agree with it. Further Solid-OIDC, SAI, and any other future spec could depend on minimalistic WebID/Profile spec. This spec could also introduce a directory that would link back to all the specs depending on it, a sort of list of dependents. This minimalistic spec could also provide guidance, security, and privacy considerations helping to decide which document (WebID or Profile) should be extended by another spec.

If for example in 2 years we will have Solid-GNAP which provides an alternative to Solid-OIDC, it would have a clear way of depending on WebID/Profile, in the same way, Solid-OIDC and SAI could depend already.

I think feedback from @acoburn and @justinwb as co-editors of mentioned spec would be very helpful.

**Comment by jeff-zucker** [comment]
@elf-pavlik, please raise other issues as you see fit and keep this discussion on the proposed  conformance rules.   I do agree with you and Ruben that this should be a minimalistic spec with no dependencies on other specs.  I do not agree with you that this spec should provide guidance on or a directory of the other specs.  The whole point of making it minimalistic is so that it can be completely orthogonal to those other specs and be able to work within the framework of any of them. If you wish to discuss this further, please raise a separate issue.

**Comment by elf-pavlik** [comment]
> it MUST include one or more triples using the `solid:oidcIssuer` predicate

To be specific, I believe this doesn't belong here. Instead, Solid OIDC should have a normative dependency on WebID/Profile spec and have a normative definition of extending WebID Document (defined here) with `solid:oidcIssuer` ([defined there](https://solid.github.io/solid-oidc/#oidc-issuer-discovery)).

This way future spec, eg. Solid GNAP could follow this pattern without needing of revising WebID/Profile spec to add `solid:gnapAssertionIssuer` (hypothetical) etc.

>> should we also submit `interop:hasAuthorizationAgent` here? We can think of it as AuthZ counterpart for the already included AuthN predicate.

> No, I don't believe it should be part of this spec. Our spec defines a very basic profile. We will enhance the final section "Other Predicates" to say that our spec is only one of the specs relating to profiles and that other specs may introduce further predicates and conformance rules required for their own purposes. We may name specific other specs here, or we may not.

I believe this *Other Predicates* should be a separate **living document** which can be amended as different specifications develop. It would have a list of predicates used in the WebID document and Profile document. This would be a lightweight version of something like IANA Link Relations registry https://www.iana.org/assignments/link-relations/link-relations.xhtml

**Comment by jeff-zucker** [comment]
@elf-pavlik  wrote 

> One could only honor statements that have the WebID as the subject or object.

We will address that elsewhere in the spec.  In this case yes, the WebID must be the subject of the solid:profileDocument triple.

**Comment by jeff-zucker** [comment]
@elf-pavlik 
>> it MUST include one or more triples using the solid:oidcIssuer predicate

> To be specific, I believe this doesn't belong here. 

That's an opinion we should consider.  Please raise a specific issue.

**Comment by jeff-zucker** [comment]
@elf-pavlik - or perhaps you are right that it does not belong here at all. If it is removed, then the dependency tree would go WebID Spec <- Solid Profile Spec <- OIDC/InterOp/Index/New Specs.  So my current thinking is remove `solid:oidcIssuer`, point to the WebID Spec for all conformance rules except the `solid:profileDocument` predicate.  Explain in non-normative text that Solid-OIDC is the currently most widely used identity protocol and that therefore the `solid:oidcIssuer` can be expected in the WebID document.

**Comment by smessie** [comment]
Great, I have no specific concerns. I'm a fan of the `space:storage` part. Now, it will however require 2 requests to get the storage location, but as @Otto-AA mentioned before, we could cache this, and by choosing the `<> a solid:ProfileDocument` implementation this could be prevented as well.

> When there is no specific reason otherwise, the ldp:inbox and space:storage triples MUST be written in the Solid Profile Document.

👍

**Comment by ThisIsMissEm** [comment]
@jeff-zucker could we allow the WebID Document to have "this document may include a `space:storage` predicate, however, to correctly find all pods for a WebID you would need to traverse the graph"

This would support ESS's model where the IdP + Storage work in tandem, and there's strict access on who/what can edit the WebID Document and who/what can create pods; The logic here is that an enterprise/government will likely want the Pod for your WebID to be hosted on their service and that you log in using their IdP, as to ensure that data privacy is maintained.

That is: ESS only allows specific applications to modify the WebID and to create Pods & associate them with a given WebID.

We currently have in production the following as a WebID Document:

```
@prefix foaf:  <http://xmlns.com/foaf/0.1/> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
@prefix solid: <http://www.w3.org/ns/solid/terms#> .

<https://id.inrupt.com/emeliasdktests>
        a                      foaf:Agent ;
        rdfs:seeAlso           <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/extendedProfile> ;
        <http://www.w3.org/ns/pim/space#preferencesFile>
                <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/settings/prefs.ttl> ;
        <http://www.w3.org/ns/pim/space#storage>
                <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/> ;
        solid:oidcIssuer       <https://login.inrupt.com> ;
        foaf:isPrimaryTopicOf  <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/extendedProfile> .
```

My understanding is that with this proposal, that would become: 

```
@prefix foaf:  <http://xmlns.com/foaf/0.1/> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
@prefix solid: <http://www.w3.org/ns/solid/terms#> .
@prefix space: <http://www.w3.org/ns/pim/space#> .

<https://id.inrupt.com/emeliasdktests> a foaf:Agent ;
        solid:oidcIssuer       <https://login.inrupt.com> .

<https://id.inrupt.com/emeliasdktests> a solid:SolidProfile ;
  space:storage        <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/> ;
  rdfs:seeAlso           <https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/extendedProfile> .
```

Where `https://storage.inrupt.com/e9179f88-ca8f-45e1-a1b9-88a83f61db5e/extendedProfile` is `a solid:SolidProfile`

Am I understanding this correctly? (I might also be getting the turtle formatting here wrong, I don't often write turtle by hand)

**Comment by jeff-zucker** [comment]
The conformance model as written would permit you to put the storage in the WebID document if you want.  But we still need to say that unless there are good reasons (such as you mention) it SHOULD be written in the Profile Document and not the WebID document (because we can't write to your WebID Document).  

A use case where I would want to insert my own storage statements is a local host.  If my desktop is running as a pod, I want organizing apps to be able to include it with my other pods.  I might change the ports on it so I might change the storage triples.  I can't do any of that if the storage predicates are only in the WebID document and the WebID document doesn't let me write.  So apps should read looking for storage in either place and write to the Profile document when they write and the server can do whatever it pleases.

I'm afraid that no, your turtle is not built according to the model I am proposing.  In the ESS case there are two documents the WebID Document and the Profile Document.  The WebID document is minimal and has a solid:profileDocument predicate pointing to the Profile Document.  This means that the WebID document doesn't need to follow all the rules (such as being writable by the WebID owner) that a Profile document has to follow.  Extended Profile Documents are pointed to from the Profile Document.  

Here is what an ESS WebID Document would look like in this model : 
```turtle 
<?WebID>
    a foaf:Agent ;
    solid:oidcIssuer <https://login.inrupt.com> .
    space:storage   <https://storage.inrupt.com/random-storage-name/> ;
    solid:solidProfile <https://storage.inrupt.com/random-storage-name/profile/card> .
```
The object of the last triple could be at any path and name and the document there should conform to the rules for a Solid Profile listed above i.e. be publicly readable, be writable by the WebID owner, have a mandatory pointer to a space:preferencesFile and optional pointers to rdfs:seeAlso files.  

If Inrupt wants to default to profiles not being publicly readable, they should do so using this model and put all profile details in the space:preferencesFile or in an rdfs:seeAlso file connected from there .  That way the Solid Profile Document is readable by the public but all actual information is not. All the public sees is that there is a private profile but that is not accomplished by making the Solid Profile Document itself unreadable.  The colors on the chart  above  show which parts can be public or private.
```

**Comment by ThisIsMissEm** [comment]
> I personally also agree that it may not make a lot of sense to lock the WebID owner out of modifying the WebID document but I don't see how we can be in the position to require servers to allow it.
> 
> Good point about the solid:oidcIssuer. My thought is that the Solid-OIDC spec is the normative one and that we refer to it. All of the conformance rules for the WebID document with the exception of the solid:solidProfile predicate are based on Solid-OIDC and we should be explicit about that.

Perhaps also keep in mind that [nothing in the spec](https://solidproject.org/TR/protocol#webid) says that a WebID must be a solid dataset or writable resource; it could just be a turtle file on an nginx webserver. As long as it has the predicates expected of a WebID Document, then it can be a WebID. (Though you might have trouble if you only serve it as html or json, given client implementations typically expect turtle right now)

It just happens that NSS, and subsequently CSS implemented the WebID as being served from storage (like ESS 1.x kind of did). 

The assumption that a WebID Document must be writable isn't backed up by the existing spec, so that's something were the spec left room for possibilities and we've seen drift in implementations.

**Comment by jeff-zucker** [comment]
@ThisIsMissEm - While the [WebID Specification](https://www.w3.org/2005/Incubator/webid/spec/identity/) over eight years old now, does not explicitly say that a profile must be editable, but almost the entire document is about writing to the profile so the assumption that the profile is writable did not come out of thin air. The approach in that specification of *writing* into the profile triples pointing to  rdfs:seeAlso to support privacy of information was expanded in the first attempt at a [Solid WebID Profile Spec](https://github.com/solid/solid-spec/blob/master/solid-webid-profiles.md) about three years ago which has been the defacto standard for app developers up until now.  

So, while this current spec supports the view that the WebID Document does not need to be directly writable via the Solid protocol, it also strongly supports the view that the Profile Document must be and that he document should follow the system of preferences and seeAlsos which have been the accepted practice for years now.

**Comment by jeff-zucker** [comment]
> It just happens that NSS, and subsequently CSS implemented the WebID as being served from storage

I don't know the history of those particular decisions, but I do know that the structure of the profile, the seeAlso files, the preferencesFile did not "just happen".  There were extensive discussions over years as shown in some of the documents I cited above.

**Comment by angelo-v** [comment]
Hi Jeff and all that are participating in that. It is a very important work.
Thank you!

After reading through it, I am having a couple of remarks and questions. I did
not read all the comments, please forgive me if I am mentioning something, that
has already been discussed. I assume the current state of discussion is
reflected and continuously updated in the initial post.

I do not get, why we need to differ between `solid:ProfileDocument` and
`foaf:PersonalProfileDocument`. Whether I can read/write a
`foaf:PersonalProfileDocument` via Solid Protocol can be determined by checking
`accept-patch` and `wac-allow` headers, can't it?

## The Solid WebID Document

> there MUST be exactly one Solid WebID Document for a given WebID URI

Why that? And which of the two documents do you mean here?

> it MUST be able to be dereferenced as RDF from the WebID URI

I assume this is already covered by the WebID spec?

> it MUST be publicly readable

Why that? I can imagine private WebID (and respective documents) I am only
sharing with my family or company. This feels like an unnecessary restriction.

> it MUST include one or more triples using the solid:oidcIssuer predicate

That's a requirement of Solid OIDC and does not need to be covered here, imho.
This spec should be independent of the auth mechanism.

## The Solid Profile Document (solid:ProfileDocument)

> there MUST be exactly one Solid Profile Document for any given WebID URI

Same as above, why only one? Of course my WebID only resolves to one document,
but every document linked via `rdfs:seeAlso` could also be such a document?

> it MUST have exactly one triple in the form <> a solid:ProfileDocument

As mentioned initially, why not stick with `foaf:PersonalProfileDocument`?

> it MUST have exactly one space:preferencesFile triple

why only one? And it could be elsewhere in the extended profile, not necessarily
in the "root".

> it MUST be publicly readable

same as above

> users can choose to restrict access to their inbox, and/or storages. In that
> case, the triples for them SHOULD be placed in the Preferences Document for
> fully private resources or in a seeAlso document for restricted audiences.
> When there is no specific reason otherwise, the ldp:inbox and space:storage
> triples MUST be written in the Solid Profile Document.

I am not sure if I understand that paragraph. Perhaps it should be split down.
In the end I would assume that I can place my `ldp:inbox` and `space:storage`
anywhere in my extended profile document or preferences.

> it MAY have zero or more triples with the solid:community predicate

What is the meaning of that predicate? Do not find it in
https://www.w3.org/ns/solid/terms

## The Preferences Document

> there MUST be exactly one Preferences Document

> it MAY have zero or more triples with the rdfs:seeAlso predicate

If I can have an seeAlso-extended preferences file, why not linking multiple
preferences files in the first place? Those could then be loaded in parallel
instead of having to follow seeALso serially.

## Extended Profile Documents

> they MAY be accessible to public, restricted, or private audiences

Agreed, but the diagram below suggests that private parts of the extended
document have to be linked from the preferences document.

> if they exist, they MUST contain triples with the WebID as subject

I think it is perfectly fine when the WebID is "only" the object, e.g. like
`schema:Event schema:performer <webId>`.

**Comment by csarven** [comment]
I think the main thing that needs to be clarified in the specification is the notion of "ownership" and "control", and if they are equivalent to, wilfully violating, or extending terms defined in the other specs. Everything should stem from that.

Introducing the notions of different kinds of Profile Documents comes across rather cumbersome at the moment, but that may just be how they are described/expressed.

Apply the principles and good practises from https://www.w3.org/2001/tag/doc/selfDescribingDocuments . (Also mentioned in https://solidproject.org/TR/protocol#self-describing-resources [SOLID-PROTOCOL].)

Some comments:

---

>[Edited to remove type indexes and reorient for wider audience]

>Given that IdPs

What's the definition of "IdP"? What kind of IdPs? What's the relationship between the notion of IdP and the WebID Profile spec? Former a dependency on WebID Profile? Why?

>may wish to restrict how WebID Documents behave, including limiting the ability of the WebID owner to write to the document

What's the definition of "WebID owner"?

Why are the "owners" restricted to write to the document?

Who can write to the document?

Who has the ultimate authority of a WebID URI?

What's the relationship between the WebID "owner" and https://www.w3.org/TR/webarch/#uri-ownership [AWWW]?

As per https://www.w3.org/2005/Incubator/webid/spec/identity/ [WebID]:

>The WebID HTTP URI must be one that dereferences to a document the user controls.

If a user may not ultimately "own" the WebID URI and/or "control" the dereferenced document (as per WebID, AWWW), then what are the differences in Solid WebID Profile?

---

>we need to support the ESS-like model in which the WebID document is not the same document as the Profile document. 

What is a "WebID document" and a "Profile document"?

>If we proceed as I've outlined below, we would have two shapes which can occur in one document or in two documents.  In the NSS-like instance where the same document is both the WebID document and the Profile document, it would need to conform to both shapes. In the ESS-like instance the WebID Document could be very minimal and the Solid Profile Document would conform to the same shape as all other Solid Profile Documents.

>I also propose two new terms a `solid:ProfileDocument` class and a `solid:profileDocument` predicate.  We can add a note that, for backward compatibility, developers may wish to also check for the foaf:primaryTopicOf and foaf:PersonalProfileDocument.

What's the definition of `solid:ProfileDocument` and `solid:profileDocument` (and relationship to `foaf:primaryTopicOf` and `foaf:PersonalProfileDocument`)?

---

A general comment on the requirement level keywords: they may not be needed, e.g. "MAY have zero or more" can simply be "have zero or more".

---


>## Solid Profile Data Model Conformance

>### The Solid WebID Document

>* there MUST be exactly one Solid WebID Document for a given WebID URI

Use a different term than "exactly" and see whether it needs to be said? What happens if there is 0 or more than 1?

>* it MUST be able to be dereferenced as RDF from the WebID URI

Already defined in [WebID]:

>A WebID Profile is an RDF document

* it MUST be publicly readable

Okay.

* it MUST include one or more triples using the `solid:oidcIssuer` predicate

The WebID owner may not necessarily use an OIDC Issuer. It may be more flexible to express this in terms of when the WebID owner uses an OpenID provider.

* it MUST either label itself as a Solid Profile Document using a triple equivalent to `<> a solid:ProifleDocument.` or include a pointer to a Solid Profile Document using a triple equivalent to `<?WebID> solid:profileDocument <?SolidProfileDocument>.`.

Express in terms of "type" or "kind" instead of "label".

>See also : [WebID Specification](TBD).

>### The Solid Profile Document (`solid:ProfileDocument`)

>* there MUST be exactly one Solid Profile Document for any given WebID URI
>* it MUST have exactly one triple in the form `<> a solid:ProfileDocument`

Same comments as earlier.

>* it MUST have exactly one `space:preferencesFile` triple
>* it MUST be writable by the WebID owner or those they designate

Clarify delegation. Perhaps `acl:delegate`.

>* it MUST be publicly readable

Okay... so all Solid Profile Documents are read-only?

>* it SHOULD have exactly one `ldp:inbox` triple
>* it SHOULD have one or more `space:storage` triples

Perhaps to be expressed similar to `solid:oidcIssuer` in that "when WebID owner wnats to.."

>* users can choose to restrict access to their inbox, and/or storages.  In that case, the triples for them SHOULD be placed in the Preferences Document for fully private resources or in a seeAlso document for restricted audiences.  When there is no specific reason otherwise, the `ldp:inbox` and `space:storage` triples MUST be written in the Solid Profile Document.

Restricting access to inbox/storage need not be said. That is a given (as with any resource). Perhaps this needs to be clarified in terms of visibility, observability, discoverability, linkability, authorization...?

Perhaps just the wording but it seems to create an either or situation. But it should be possible for WebID owner to do all of those things at the same time.

* it MAY have zero or more triples with the `solid:community` predicate
* it MAY have zero or more triples with the `rdfs:seeAlso` predicate

Okay..

* it MAY be the same document as the WebID Document or a separate Document. In the case in which the Solid Profile Document and the Solid WebID Document are the same, the document MUST conform to the rules listed above for both types of documents.

This is confusing.

### The Preferences Document (`space:preferencesFile`)

* there MUST be exactly one Preferences Document
* it MUST be writable by the WebID owner or those they designate

This is a given. Perhaps the assumption that any resources that's being described is/may be subject to authorization should be stated earlier so that these statements need not be said.

* it MUST NOT be readable or writable by anyone other than the WebID owner or those they designate

Need not be said? Doesn't that essentially cover all possibilities?

* it MAY have zero or more triples with the `rdfs:seeAlso` predicate

Okay..

### Extended Profile Documents (`rdfs:seeAlso`)

* there MAY be zero of more extended profile documents

Okay..

* they MAY be accessible to public, restricted, or private audiences

Need not be said?

* if they exist, they MUST contain triples with the WebID as subject

Okay. But perhaps not solely.

**Comment by RubenVerborgh** [comment]
> > * there MUST be exactly one Solid WebID Document for a given WebID URI
> 
> Use a different term than "exactly" and see whether it needs to be said? What happens if there is 0 or more than 1?

I think it's even worse; his is non-enforceable. I cannot stop anyone from creating a document for any given WebID URI.

Can we rephrase this sentence as to who is expected to (not do) what?

**Comment by jeff-zucker** [comment]
> Why are the "owners" restricted to write to the document?

I should clarify - owners can write to a WebID document even in ESS.  But they have to use a server-script that does not follow the Solid Protocol.  So they still own it in the sense of being  able to write to it, just not via a Solid app.

> Who has the ultimate authority of a WebID URI?

Issuers and users have different parts of the authority.  Do we need to define that here?

> What's the definition of "WebID owner"?
> Who can write to the document?
> What's the relationship between the WebID "owner" and https://www.w3.org/TR/webarch/#uri-ownership [AWWW]?

These issues will all need to be addressed.  The text here is not meant to be fully explanatory, rather  to spark discussion.  We definitely need to define (or in most cases refer to other specs' definitions for) the ideas of WebID, ownership, control, and delegation.  The section on the WebID Document will refer to the WebID spec and not change or require anything other than the solid:profileDocument predicate.

> As per https://www.w3.org/2005/Incubator/webid/spec/identity/ [WebID]:
> 
> > The WebID HTTP URI must be one that dereferences to a document the user controls.
> 
> If a user may not ultimately "own" the WebID URI and/or "control" the dereferenced document (as per WebID, AWWW), then what are the differences in Solid WebID Profile?

A user can control a WebID Document in ESS using server-provided access that does not follow the Solid Protocol.  So the user does have control but a Solid app operating on their behalf does not because it is not on a Solid server and does not need to conform to the Solid Protocol.  So I don't think our spec adds any special rules or requirements on top of the WebID Spec other than the solid:solidProfileDocument triple.  

> What is a "WebID document" and a "Profile document"?

If you can't answer that from the conformance rules, I'm at a loss to explain, let's chat if you are still unclear on this.

> What's the definition of ...  `solid:profileDocument` (and relationship to `foaf:primaryTopicOf` )?

A predicate whose subject is a WebID and whose object is a Solid Profile Document as defined in the rules here listed.
 
> A general comment on the requirement level keywords: they may not be needed, e.g. "MAY have zero or more" can simply be "have zero or more". ... Use a different term than "exactly" and see whether it needs to be said? What happens if there is 0 or more than 1?

Noted!  Will drop the unneeded MAYs and the exactlys.

> Already defined in [WebID]:

Yep, already edited to point there.

> The WebID owner may not necessarily use an OIDC Issuer. 

Right, edited to remove this from the rules.

> > * it MUST be writable by the WebID owner or those they designate
> 
> Clarify delegation. Perhaps `acl:delegate`.

TBD.

> > * it MUST be publicly readable
> 
> Okay... so all Solid Profile Documents are read-only?

No, they must be writable by their owners and readable by the public.  Any other permissions or lack of them are permitted.

> > * it SHOULD have exactly one `ldp:inbox` triple
> > * it SHOULD have one or more `space:storage` triples
> 
> Perhaps to be expressed similar to `solid:oidcIssuer` in that "when WebID owner wnats to.."

Yes, good idea.

> Restricting access to inbox/storage need not be said. That is a given (as with any resource). Perhaps this needs to be clarified in terms of visibility, observability, discoverability, linkability, authorization...?

Yes, good point, it's not about whether the inbox is private it's about whether the triple describing the location of the inbox is private.

> > * it MAY be the same document as the WebID Document or a separate Document. In the case in which the Solid Profile Document and the Solid WebID Document are the same, the document MUST conform to the rules listed above for both types of documents.
> 
> This is confusing.

I'm open to better way to say it.  If the WebID dereferences directly to this document, it needs to follow the rules of both documents.  If the WebID dereferences to a WebID Document that points to this document, this document only needs to conform to the rules for a Solid Profile Document but not the rules for both documents.

> > * it MUST be writable by the WebID owner or those they designate
> 
> This is a given. 

Um, yep, I should remove that line.

> > * it MUST NOT be readable or writable by anyone other than the WebID owner or those they designate
> 
> Need not be said? Doesn't that essentially cover all possibilities?

It needs to be said for app developers who are creating a preferencesFile when none exists - they need to know the access rules to place on it.

> > * they MAY be accessible to public, restricted, or private audiences
> 
> Need not be said?

Perhaps move it to a note.

> > * if they exist, they MUST contain triples with the WebID as subject
> 
> Okay. But perhaps not solely.

Right.  I edited the top post to reflect this.

**Comment by jeff-zucker** [comment]
> Can we rephrase this sentence as to who is expected to (not do) what?

We will have prior text describing that this spec is aimed at app developers; that the conformance rules are meant to assist them in learning what to expect in a well-formed profile; and how to create a well-formed profile if they are an app able to do that.  Since servers vary, and users can edit, delete, and change access on any resource, there is no guarantee that an app will have access to a well-formed profile.  In that case, fixer apps can offer to make the profile well-formed and can do so by following the rules listed here.

**Comment by RubenVerborgh** [comment]
@jeff-zucker Let me rephrase.

---

For example, here's a sentence from the Solid Protocol spec:

> Clients MUST use the Content-Type HTTP header in PUT, POST and PATCH requests

Note that there is an agent ("client") as the subject of the sentence.

Imagine a rewriting of the sentence with a subject without agency ("request"):

> PUT, POST and PATCH requests MUST use the Content-Type HTTP header.

That sentence can be interpreted in many different ways:
- Servers MUST error if the `Content-Type` header is missing
- Servers MUST add the `Content-Type` header if it is missing
- Servers MUST ignore requests without `Content-Type` header
- Clients MUST specify a `Content-Type` header
- Libraries MUST specify a `Content-Type` header
- Apps MUST specify a `Content-Type` header
- …

What makes the sentence enforceable is the usage of an agent.

---

So what I mean is that

> there MUST be exactly one Solid WebID Document for a given WebID URI

is non-actionable.

Who MUST do what?

Because I can trivially write a Web service (and don't dare me 😄) that invalidates that statement for all current and future WebIDs.

---

> there is no guarantee that an app will have access to a well-formed profile

Ack.

> fixer apps can offer to make the profile well-formed

That specific MUST, they can never implement. It's an impossibility due to the way it is written.

**Comment by RubenVerborgh** [comment]
> > That specific MUST, they can never implement. It's an impossibility due to the way it is written.
> 
> Sorry, which MUST?

I mean to say that

> there MUST be exactly one Solid WebID Document for a given WebID URI

Is entirely unimplementable. There's nothing any agent in the ecosystem could do to make it true.

And I don't mean to just pick on that statement; others have similar problems because of the lack of an agent.

**Comment by jeff-zucker** [comment]
@RubenVerborgh thanks, didn't know if you meant an additional point.  

All requirements will have agents. The description of the Solid WebID Document will be left to the WebID spec with the exception of some non-normative text and the requirement for the solid:solidProfile predicate.

**Comment by NoelDeMartin** [comment]
I just went through the entire discussion, thanks @jeff-zucker and everyone else involved in the efforts thus far :).

Here's my five cents:

> the WebID owner or those they designate

This is mentioned in a couple of places, but I'm not sure what it means.

Since it's written with an OR, instead of an AND, does it mean that it would be possible that I am the owner of a WebID and I'm not able to write in my own profile?

> it MUST be publicly readable [Note : we're not sure, thoughts?]

Yes, I think it should be publicly readable.

I've read @angelo-v's comment on private WebIDs, but I think in that case the WebID Document and Profile Document should only contain the bare minimum to make them valid. All additional information can be in the Restricted Extended Profile.

> All of which yields this updated diagram

Looking at that diagram, I understand how it works but I think some of the names are confusing. Why is one called "Preferences Document" and then "Private Extended Profile Document(s)"? Wouldn't it be possible to call everything Profile Document(s)? Then we could have:

- WebID Document
- Public Profile Document
- Private Profile Document
- Public Extended Profile Document(s)
- Restricted Extended Profile Document(s)
- Private Extended Profile Document(s)

I guess in this case, it'd make more sense to use `solid:privateProfile` instead of `space:preferencesFile` (and we can still mention it for backwards compatibility as we're doing with `foaf:PersonalProfileDocument`).

> [@ThisIsMissEm] Perhaps also keep in mind that nothing in the spec says that a WebID must be a solid dataset or writable resource; it could just be a turtle file on an nginx webserver.

This makes me wonder, in that case, I suppose the WebID Document will not respond to `Accept` headers for things like JSON-LD, right?

I suppose in that case, it should be explicitly noted that WebIDs only successfully return Turtle. All other RDF formats should just be up to each server's implementation.

Looking at [the Solid spec](https://solidproject.org/TR/protocol#resource-representations), it seems like at least JSON-LD is mandatory for Solid documents.

> [@jeff-zucker] there is no guarantee that an app will have access to a well-formed profile. In that case, fixer apps can offer to make the profile well-formed and can do so by following the rules listed here.

I'm not sure what you mean with this. I know that in practice there could be malformed profiles in the wild. But in that case, aren't they violating the spec and it's not an app's responsability to fix them?

I guess you're just mentioning it because it would be nice if those fixer apps could exist, so it's important for the spec to allow for that. But I guess it's to be expected that apps don't work properly if there is something wrong with the profile, right?

---

There were also a couple of mentions about caching in order to mitigate the issue of making many requests (@Otto-AA, @smessie).

I've recently implemented this in an app, and there is a problem that maybe could be addressed by this spec (should I open an issue to discuss it?).

The gist of the problem is that caching the profile is great, but then we're faced with the issue of deciding when to invalidate this cache. For things like an avatar or display username it may not be too important, but when it comes to adding things on a type index or additional storages, it's important that apps are aware of the new information.

I was thinking on adding some advanced setting in my app to force it, but it would be great if the Profile Document included a triple indicating the last time anything in the profile was modified. That way, I could just do one request and know if I need to load the entire chain of documents again.

This is similar to an ongoing discussion in the main Solid spec: https://github.com/solid/specification/issues/227#issuecomment-773945439

**Comment by jeff-zucker** [comment]
> > the WebID owner or those they designate
> 
> This is mentioned in a couple of places, but I'm not sure what it means.

Definitely I was unclear. We will specify this much more completely and the word I meant was delegate.  We will point to a spec (acl I believe) that defines delegation.  Basically it covers situations such as a child delegating ownership to a parent.

> Looking at that diagram, I understand how it works but I think some of the names are confusing. Why is one called "Preferences Document" and then "Private Extended Profile Document(s)"? Wouldn't it be possible to call everything Profile Document(s)? Then we could have:
> 
> * WebID Document
> * Public Profile Document
> * Private Profile Document
> * Public Extended Profile Document(s)
> * Restricted Extended Profile Document(s)
> * Private Extended Profile Document(s)
> 
> I guess in this case, it'd make more sense to use `solid:privateProfile` instead of `space:preferencesFile` (and we can still mention it for backwards compatibility as we're doing with `foaf:PersonalProfileDocument`).

Hmm, I like that a lot.  We'll discuss.

> 
> > [@ThisIsMissEm] Perhaps also keep in mind that nothing in the spec says that a WebID must be a solid dataset or writable resource; it could just be a turtle file on an nginx webserver.
> 
> This makes me wonder, in that case, I suppose the WebID Document will not respond to `Accept` headers for things like JSON-LD, right? I suppose in that case, it should be explicitly noted that WebIDs only successfully return Turtle. All other RDF formats should just be up to each server's implementation.
> Looking at [the Solid spec](https://solidproject.org/TR/protocol#resource-representations), it seems like at least JSON-LD is mandatory for Solid documents.

Yes, but the WebID Document doesn't have to follow the Solid spec - it is not necessarily on a Solid resource server.  The Solid Specs are not relevant to all WebID Documents, That said, the WebID spec says turtle is a MUST and other formats are a MAY so yes, we should warn developers.
 
> > [@jeff-zucker] there is no guarantee that an app will have access to a well-formed profile. In that case, fixer apps can offer to make the profile well-formed and can do so by following the rules listed here.
> 
> I'm not sure what you mean with this. I know that in practice there could be malformed profiles in the wild. But in that case, aren't they violating the spec and it's not an app's responsability to fix them?

It is not an app's responsibility to fix a profile but neither is it anyone elses.  There is no spec telling servers that they must create profiles in a given manner and @timbl has been very clear that this spec should not be aimed at servers and that apps ARE the only part of the ecosystem that can ensure valid profiles.  This is why he and Timea have been changing the discovery code in SolidOS to *write* missing parts of the profile.
 
> I guess you're just mentioning it because it would be nice if those fixer apps could exist, so it's important for the spec to allow for that. But I guess it's to be expected that apps don't work properly if there is something wrong with the profile, right?

They will exist as soon as the latest mashlib is merged. :-)  And yes, it's an important part of the spec to not only say apps can do that but to also describe how to do it if able and willing.  And yes, if the one of the pieces of a Profile mentioned in the spec is missing, the profile can be considered broken.  Apps might be able to get what they need from even a broken profile (depending on how it is broken) but there is no guarantee..                                                    

> 
> There were also a couple of mentions about caching in order to mitigate the issue of making many requests (@Otto-AA, @smessie).
> 
> I've recently implemented this in an app, and there is a problem that maybe could be addressed by this spec (should I open an issue to discuss it?).

Yes please.

> The gist of the problem is that caching the profile is great, but then we're faced with the issue of deciding when to invalidate this cache. For things like an avatar or display username it may not be too important, but when it comes to adding things on a type index or additional storages, it's important that apps are aware of the new information.
> 
> I was thinking on adding some advanced setting in my app to force it, but it would be great if the Profile Document included a triple indicating the last time anything in the profile was modified. That way, I could just do one request and know if I need to load the entire chain of documents again.
> 
> This is similar to an ongoing discussion in the main Solid spec: [solid/specification#227 (comment)](https://github.com/solid/specification/issues/227#issuecomment-773945439)

Interesting idea, please raise a separate issue and we'll discuss.

**Comment by angelo-v** [comment]
> > it MUST be publicly readable [Note : we're not sure, thoughts?]
> 
> Yes, I think it should be publicly readable.
> 
> I've read @angelo-v's comment on private WebIDs, but I think in that case the WebID Document and Profile Document should only contain the bare minimum to make them valid. All additional information can be in the Restricted Extended Profile.

What is the point in having a public profile, that does not contain any public data? And why should we enforce organizations to publish all their employees WebIDs? Why is it important that WebID documents are publically readable? Even WebIDs that are only reachable within an intranet should work quite fine.

**Comment by jeff-zucker** [comment]
The Editors are hotly debating the topic of the privacy of the Solid Profile Document (the main profile document, not all the others), so thanks much @angelo-v and @NoelDeMartin for chiming in.  We will previously have defined "private" as accessible only by the WebID owner.

>  why should we enforce organizations to publish all their employees WebIDs? 

Whoa, how does that follow from having the Solid Profile Document be publicly readable?  We are NOT talking about making the entire set of profile documents readable, only the main Solid Profile Document. 
The employee WebIDs belong in a seeAlso document restricted to HR or to whomever the organization decides, they should never go in the main Profile Document.  

And this illustrates the answer as to why there should be a public main Profile Document with no public data.  If the organization puts the WebIDs in a restricted seeAlso file and the main Profile Document only contains a pointer to that file and the organization later decides they  want to make some information public, they can put the public information in the previously almost empty Profile Document and leave the employee WebIDs in the protected document and they are done.  If they start by putting the WebIDs in a private main Profile Document and then want to present some public information, they'd need to move the private material out of the Profile Document, create and create access rules for a different document, change the access rules for the Profile Document.   Why not start with proper structure and put restricted things in places meant to be restricted?

**Comment by jeff-zucker** [comment]
I guess the case of an organization that *only* ever wants to present information to its members and will never have any public access is somewhat different.  I still argue that it should follow the principles of putting restricted materials in restricted seeAlso files - that way apps will be able to look for the same structure for that organization as everywhere else.

**Comment by jeff-zucker** [comment]
I especially am against servers "protecting" their users by defaulting to a private main Profile Document - this would leave the user with a very messy task if they later decide to make parts of it public.  If the server wants to default to a private Profile (not Profile Document) it should  do so by putting all information in  private seeAlso file.  I am all in favor of defaulting to a private Profile.  I am against defaulting to a private Profile Document.

**Comment by jeff-zucker** [comment]
Users should have a very clear picture - if I put this thing here, it is Public, if I put it there it is Private, and if I put it in a place I restrict, it is Restricted.   They should know - if I put something in the main Solid Profile, it is Public, so if I want it Restricted or Private, I need to put it somewhere else.

**Comment by jeff-zucker** [comment]
The proposal we're making here aims to make it crystal clear where different kinds of resources belong

* Private resources belongt in the Preferences Document or an Extended Document pointed from there
* Restricted resources belong in Extended Documents pointed to from the main Profile Document
* Public resources belong in the main Profile Document or in an Extended Document

**Comment by jeff-zucker** [comment]
I phrased the previous comment as coming from the Editors.  It is my opinion, only.  Some other Editors are not in favor of recommending a public main Profile Document.  Nothing is written in stone or even invisible ink.

**Comment by RubenVerborgh** [comment]
> The Editors are hotly debating the topic of the privacy of the Solid Profile Document

I want to state for the record (because my comment is not actionable yet) that strong assumptions / coupling is being made here between the document API and the permissioning system.

I think that, fundamentally, dividing information between public and private is flawed, and that we are basically hitting the limits of document-orientation and access control. Simply said, for most decisions, counterexamples can easily be found.

A simple but crucial pet peeve is a piece of information as simple as a person's name.
I want my banking app to see my given and family name. I want my dating app to only see my given name.
So which information goes in my profile document?

And this argument holds in general: the whole design is starting to be based on a permissions model, whereas permissions are supposed to be orthogonal in Solid.

**Comment by jeff-zucker** [comment]
> I think that, fundamentally, dividing information between public and private is flawed

I have to wonder then how Solid in general and this spec in particular can address the goal that @timbl has set stated : "provide a gradient of intimacy between the private and the public".

> we are basically hitting the limits of document-orientation and access control

There are certainly going to be different solutions in the future, but documents and access control are what we have now.  

> A simple but crucial pet peeve is a piece of information as simple as a person's name.
> I want my banking app to see my given and family name. I want my dating app to only see my given name.
> So which information goes in my profile document?

Your family name triple would go in a see:Also document permissioned to only allow your banking app and others you want to know it.  Your given name triple would go in the main profile document if you want everyone to know it, or, if that is also for a restricted audience, in a see:Also document permissioned to allow only your banking app, your dating app, and others you want to know it. In that latter case, neither triple would be in the main profile document.

**Comment by RubenVerborgh** [comment]
> I have to wonder then how Solid in general […] can address […] "provide a gradient of intimacy between the private and the public".

Access control. We do this through access control already.

The problem here is that this access control information is being replicated (and incompletely) in another system.
Type indexes have a similar problem.

> Your family name triple would go in a see:Also document permissioned to only allow your banking app and others you want to know it. Your given name triple would go in the main profile document if you want everyone to know it, or, if that is also for a restricted audience, in a see:Also document permissioned to allow only your banking app, your dating app, and others you want to know it. In that latter case, neither triple would be in the main profile document.

That's not going to work, because then I'd need to know beforehand what apps I will be using.

It also shows that things are context-dependent. I'm happy with you knowing my first and last name if you're a person visiting my website. But I'm not happy for any random website I visit to know my name. So the "everyone" in "everyone can know my first and last name" is context-dependent.

But this goes to show that we should not have other systems interfere with the access control system. Access control will evolve to tackle some of the issues above, and having another system contain a loose (public/private/restriction) replication of it is therefore guaranteed to cause problems.

I don't think we can solve this here though; I'm just listing it for the record. It's an inherent issue with the LDP/ACL abstraction. What we can do know is not dig ourselves deeper into it by creating another extra tie to ACL.

**Comment by jeff-zucker** [comment]
@RubenVerborgh lots of food for thought in your posts.  May I ask if you see any value at all in having a Profile spec?  What would it clarify if it can't clarify the WebID Document (because it might not be on a Solid resource server) and it can't clarify permissions (if we accept your reasoning), what would it say?  Go to the WebID URI and keep opening things that say they are profile documents until you find what you need.  I'm not being rhetorical, is that all we can tell developers?

**Comment by VirginiaBalseiro** [comment]
I will try my best to summarize my views without repeating stuff already said above. 

I agree with Ruben that this spec should not revolve around how access control is applied to any resources.

I also see, as an app developer, the difficulty in designing UIs that feel familiar to both users and developers, and the need to make this more predictable for developers.

I do however think that since Solid imposes a different paradigm, we should strive to make this transition easier, but still setting the correct expectations.

The way I see the Solid ecosystem in a future is, servers (and therefore different providers, identity providers, Pod providers, and so on) will offer different features and users will prefer one or the other or many based on their own preferences. This ties into the idea of the user having choice over what to do with their online identity and personal data. 

Apps should be prepared for a scenario where a WebID document is not editable and contains no links to any profile document. Would the app be unusable? In some cases, yes. Depending on the user case, it will likely be limited in what it can do unless the user contacts their provider and voices their concern, or changes provider. The audience of this spec is app developers, not server implementers. 

What I am saying is that I think this document should not make limiting requirements such as there must be exactly one profile… or the profile must be private/public... I think we need to describe what the possibilities are, and recommend a reasonable course of action.

To take the above question by Jeff as an example, what should an app do if there is a huge number of profiles, should it follow all… I would recommend always asking the user (and this applies to multiple anything, like multiple Pods). “We found X amount of profiles, pick one for this thing you are trying to do/using this app”. There are examples of this, just speaking from the user interaction perspective, one being Google profiles. 

So all in all, it’s not that I don’t see the advantage in trying to set a limit to these things, it is more that I don’t think it is this spec’s responsibility to do so, and I also think it might be better to focus our efforts on trying to reach a set of reasonable best practices around designing UIs. This is a big challenge because of the paradigm shift, so I believe the more resources we have tackling this topic (such as this spec), the better. 

So I would say: you can expect that either the WebID document is editable, or it contains a link to an extended profile document, or many. Prefer those which are a `solid:ProfileDocument` (or linked through `solid:profileDocument`), but understand that there might be others, and weigh the pros and cons of fetching them all depending on your app’s needs. Be prepared for the case where neither is true, and you can only work with a WebID and take it into account when handling errors in your application.

**Comment by jeff-zucker** [comment]
@VirginiaBalseiro  wrote

> Apps should be prepared for a scenario where a WebID document is not editable and contains no links to any profile document. 

I agree.  Our spec can't force anything about the WebID document.  I retract anything I've said so far that implies that we should. 

> I would recommend always asking the user (and this applies to multiple anything, like multiple Pods)

A profile isn't necessarily used by the owner so what would a visiting app run by someone other than owner do?  What would be offered to them to choose from?  What about bot apps that just want to read or write something?  Who would be asked to choose?  

>  you can expect that either the WebID document is editable, or it contains a link to an extended profile document, or many

Is the app responsible for opening extended profile documents inside extended profile documents?  If so, when does that stop? 

If the WebID document is not editable by the Solid Protocol (which is its right), and there can be many extended profile documents, but no single Solid Protocol addressable location to point to them, how does an app create a link to one of these extended profile documents other than by creating a link to an extended profile document inside another extended profile document?

**Comment by jeff-zucker** [comment]
@VirginiaBalseiro wrote : 

  > I agree with Ruben that this spec should not revolve around how access control is applied to any resources.

I agree with that also.  The recommendations I would like to see are about *if* access controls are applied, not *how*.  There are really only two questions my recommendation deals with and neither depends on how permissions are handled or restricts users ability to handle differing consents.
```
is the user of the app the owner of the profile being visited?                                                                                        
  yes -> should the information be available to all apps visiting this profile?                                                                       
    yes : preferences file                                                                                                                            
    no  : extended document pointed to from the preferences file                                                                                      
  no ->  should the information be available to all users and all apps visiting this profile?                                                         
    yes : profile document                                                                                                                            
    no  : extended document pointed to from the profile document                                                                                      
```

**Comment by jeff-zucker** [comment]
@VirginiaBalseiro  wrote

> this document should not make limiting requirements such as ... the profile must be private/public

I will now agree with this point. The profile document could be private/pubic/restricted as owner wants.  There would hopefully be UIs to easily convert the permissions.

**Comment by jeff-zucker** [comment]
> you can expect that ... the WebID document is editable

No, you can't, but a WebID document and a Solid Profile Document are not the same thing.  You should, I believe, be able to expect that a Solid Profile Document is editable by the WebID owner using a Solid software of their own choosing.  The idea that the owner would need to contact the Identify Provider in order to edit their own profile makes the idea of ownership meaningless.  So if the WebID Document is not editable using Solid software, it is free to not point to a Profile Document or anything else it wants.  But if it wants apps to interoperably connect to a Solid Profile, it needs, in my opinion, to point to a document or documents the WebID owner can edit using the Solid protocol.

**Comment by jeff-zucker** [comment]
^^ For example, it seems to me to be completely against the Solid philosophy that the Identity Provider, without my consent, makes my storages visible to the world and does not allow Solid apps I choose to add, remove, or change the visibility of storages.  Those things, to me mean I do not own my own profile.

**Comment by NoelDeMartin** [comment]
Thanks @jeff, I opened a follow up about the caching in https://github.com/solid/webid-profile/issues/47

> [@angelo-v] What is the point in having a public profile, that does not contain any public data?

> [@RubenVerborgh] the whole design is starting to be based on a permissions model, whereas permissions are supposed to be orthogonal in Solid.

> [@VirginiaBalseiro] I agree with Ruben that this spec should not revolve around how access control is applied to any resources.

Although I agree with the general idea that permissions shouldn't be tackled heavily in this spec, I still think it's important to set clear guidelines for what apps can expect from a WebID.

In particular, I want to at least know whether a url is a WebID or a random url from the internet. If a WebID is not publicly-readable and I get an error response, I don't know if that's because the WebID is private or it just isn't a WebID.

This concern arises from the UX difficulty [we discussed in the forum](https://forum.solidproject.org/t/has-any-work-been-done-to-standardise-ux-patterns-for-logging-into-solid-applications/5251/2?u=noeldemartin). TLDR: I can't be sure if what users introduce when logging in is a WebID or an issue provider url or what. So in my apps, I try to search for a valid WebID and get the issue provider from there. If I can't, I just use whatever they typed as the issue provider.

Having that into account, I think it's useful that WebIDs are public even if the only public information is that they are indeed a WebID and what issue provider to use.

> [@VirginiaBalseiro] To take the above question by Jeff as an example, what should an app do if there is a huge number of profiles, should it follow all… I would recommend always asking the user (and this applies to multiple anything, like multiple Pods). “We found X amount of profiles, pick one for this thing you are trying to do/using this app”

As I'm understanding this spec, all profile documents are conceptually part of the same profile, they are just separated in order to have different permissions, reduce the number of requests, or any other reasons.

So the idea of querying all documents or not is just an implementation detail to improve performance, but I don't think it's a problem a user should be concerned about.

**Comment by jeff-zucker** [comment]
> I still think it's important to set clear guidelines for what apps can expect from a WebID.
> In particular, I want to at least know whether a url is a WebID or a random url from the internet. 

The WebID Document is defined in the WebID spec which says in a non-normative section "A WebID Profile is a public document".  The Solid-OIDC spec raises that as an issue but does not definitively say that a Solid-OIDC compliant WebID Document must be public.  If you are interested in ensuring a public WebID document you should contribute to that spec : https://solidproject.org/TR/oidc#oidc-issuer-discovery or to the WebID spec.

So the question of whether the WebID Document is public will not be decided in this current spec.  What we will have to decide on is whether, when the WebID Document is different from the Solid Profile Document and only points to that document whether that pointed to document, the Solid Profile Document, is public.  The consensus seems to be that we should not specify that.

**Comment by jeff-zucker** [comment]
@NoelDeMartin here are the definitions I would like to  use : 


From [Solid Protocol](https://solidproject.org/TR/protocol#solid-app) : "A `Solid app` is an application that reads or writes data from one or more storages."

From the [WebID Specification]() : "A `WebID Profile Document` is an RDF document which uniquely describes the Agent denoted by the WebID in relation to that WebID."

Proposed by me : "A `Solid Profile` is a set of WebID Profile Documents that are readable and writable by properly authorized Solid apps and whose starting point can be dereferenced either directly from the WebID URI or indirectly by finding a solid:profileDocument triple in the document the WebID URI dereferences to."

Adapted by me from [Solid Protocol](https://solidproject.org/TR/protocol#owner) - "A `WebID owner` is a person or a social entity that is considered to have the rights and responsibilities of a WebID and its associated Solid Profile. An owner is identified by a URI, and implicitly has control over all data in the Solid Profile associated with that URI."

**Comment by timea-solid** [comment]
This discussion clearly shows the hard line the editors need to walk between _What is possible now with the Solid Profile_ and _What we want the Solid Profile to be in the future_.  Initially, we set out to write down _how it is right now_ so that dev have an easier time just focusing on what they love: code. 

As a result, we should take a step back and leave some things as they are -> meaning: link to the location in the other specs where you find more information. For example link to: "The Solid-OIDC spec raises that as an issue but does not definitively say that a Solid-OIDC compliant WebID Document must be public." This is something we already realized we need to do in more places of this spec. We can take this discussion point in the weekly meetings because it touches on the goal of the spec. 

As for the topic itself: my wish is to have in the future a so-called triple level privacy control, a case in which you can have only one Solid Profile which can have all info and control the user wants/needs.  The 'possibility' should stay open after this spec.  (As an example: we decoupled type indexes from WebID to make sure we don't stop progress on Solid indexes.)

**Comment by jeff-zucker** [comment]
@timea-solid - I agree that we should, as much as possible, depend on existing specs and not duplicate their work.  We should talk about the WebID Profile Document and OIDC non-normatively as part of the discovery process but not make any normative statements about them.   

I also agree that the future is permissioned triples and that our spec absolutely should not do anything that would impede the development or deployment of permissioned triples or other future developments.  OTOH, when permissioned triples are available, this spec will need to be changed.  We can't possibly write a spec for permissioned triples now.  And when will permissioned triples be available?  (one year? five years?).  I think it would be a disaster to have no profile spec between now and then.  Is document permission messy and should be replaced ? Yes, definitely.  Is it what we have to work with today?  Yes.

To me the most important parts about the profile are 

1) the Solid Profile should be modifiable by solid apps of the user's choosing and totally under the WebID owner's control (see [my proposed definitions](https://github.com/solid/webid-profile/issues/40#issuecomment-1236391633)) .  The Identity Provider is welcome to maintain control over  the document the WebID URI dereferences to for their own purposes (such as oidcIssuer) when it is located on their own servers rather than on a Resource Server, but it can not claim that that document is a Solid Profile.  It isn't, it points to a Solid Profile.  Otherwise apps can not modify profiles and apps are the point in the ecosystem that need to modify profiles.  Today we have a server putting our storages in public view without our permission, what is to stop a server from putting our name and passport number in the WebID document and make us make requests to them to change it?  If servers rather than apps determine the shape of profiles and we can't aim a specification at servers to control that process, that means profiles are irrevocably living in the wild west where each Identity Provider gets to  decide what a Solid Profile looks like.  In my opinion, the server is free to add other features, but it can not do so by interfering with a user's ownership of their own profile and all of the data in it and it. 

2)  There should be a clear way for users and app developers to know where to put and look for differently permissioned materials.  It should, as much as possible, support existing practice.   SolidOS, the Inrupt libraries, and many other apps are designed to work within the document permissioning system.  I am not in favor of telling all of those apps, oh forget about that, just put stuff where you want to.  I won't repeat myself on this, I think [this](https://github.com/solid/webid-profile/issues/40#issuecomment-1236202594) sums up my view on permissioning.

**Comment by jeff-zucker** [comment]
@timea-solid - in response to your mention of triple-level permissioning, I said "We can't possibly write a spec for permissioned triples now." but I now believe I am wrong about that.  We can say an app writing or repairing a profile on a server that supports triple-level permissioning should, when possible put all triples directly in the Solid Profile Document but may, if it wishes, follow the document model we propose.  Apps on servers that only support document level permissioning should follow the recommendations we propose for writing documents.  Apps reading a profile on either type of server should read the Solid Profile Document and if that doesn't tell them what they need to know, follow the document recommendations we make to load and query.  That seems to me to make our spec work with either kind of permissioning.

**Comment by jeff-zucker** [comment]
I believe Trinpod has already implemented triple-level permissioning.  @gibsonf1 would the paragraph above this work for you?

**Comment by csarven** [comment]
re Editor Role:

The editor is expected to apply the consensus decisions of the Group. The editor is not expected to have all the responsibilities of the author or rights to matters which the Group does not explicitly decide. In practise, there is leeway. Editors need to practise egoless writing and they can be influential voices in Group deliberations.

(I'll look into adding something along these lines to CONTRIBUTING.md.)

---

re Orthogonality:

It is expected that specifications are kept orthogonal ( https://www.w3.org/TR/webarch/#orthogonal-specs ) and overlapping features are clearly indicated: https://solidproject.org/TR/protocol#introduction .

The WebID Profile specification is deemed to focus on identity - attached data towards presenting an agent in different contexts - and discovery, without venturing too far into identification, authentication, authorization, or modifying abstractions described elsewhere.

---

re Profile

https://gitter.im/solid/webid-profile?at=6310b75fd020d223d3181b8f :

>Clarify how a profile is a presentation of an agent.
>whether an agent can have different profiles and what their relationships may be..
>"Extended profile" is a bit of plumbing (but essentially about the same identity). It is an answer for the need to distribute information into different resources for different reasons, access privileges being one.

To define:
* https://solid.github.io/webid-profile/#solid-profile
* https://solid.github.io/webid-profile/#extended-profile

To introduce, reuse, and clarify relationship:

re "owner", GOTO https://github.com/solid/webid-profile/issues/40#issuecomment-1231860197

re "agent": The SOLID-PROTOCOL defines "agent" https://solidproject.org/TR/protocol#agent as:
>An agent is a person, social entity, or software identified by a URI; e.g., a WebID denotes an agent [WEBID].

---

re Authentication

Clarify the abstractions involving "Identity Provider".

Consider the following for a moment: if it is deemed that a WebID Profile Document may not necessarily be a resource available from a Solid server, could the same kind of assumption not be made about a "provider" (or even Solid-OIDC in particular)? 5+ years ago we would a similar consideration could've been made when WebID-TLS was the focus for authentication.

---

re WebID Profile Document Access Privileges:

The intended access privileges about WebID Profile Documents is clarified in https://github.com/w3c/WebID/pull/7 ( as per https://github.com/w3c/WebID/issues/1 ) in that the WebID 1.0 specification does not set a constraint - documents can be public or protected, and they may contain public or private information.

**Comment by jeff-zucker** [comment]
> re Editor Role:
> 
> The editor is expected to apply the consensus decisions of the Group. 

Not sure why you're mentioning that here, but I've tried to speak only for myself and have clarified that several times in the discussion.  The top posting will try to reflect the group consensus, not my opinion and other Editors are free to edit it. 

> re Orthogonality:

I agree entirely.  We should not redefine anything in the  WebID or Solid-OIDC and should conform to previous specs except in the few case where we are introducing something new.

> The WebID Profile specification is deemed to focus on identity - attached data towards presenting an agent in different contexts - and discovery, without venturing too far into identification, authentication, authorization, or modifying abstractions described elsewhere.

Yes, agree, this spec should focus on presentation of identity and on discovery, and refer to other specs  for identification, authentication, and authorization.

> > Clarify how a profile is a presentation of an agent.
> > whether an agent can have different profiles and what their relationships may be..

Absolutely.  This is a very important step, I agree.  I already have a number of use case scenarios but there should also be an overview.

> To define:
> 
> * https://solid.github.io/webid-profile/#solid-profile
> * https://solid.github.io/webid-profile/#extended-profile

Yes.

> 
> To introduce, reuse, and clarify relationship:
> 
> re "owner", GOTO [#40 (comment)](https://github.com/solid/webid-profile/issues/40#issuecomment-1231860197)
> 
> re "agent": 

Yes point to Protocol and WebID for this.

> re Authentication
> 
> Clarify the abstractions involving "Identity Provider".

I think we can skip much mention of them.  But yes, if we use the term we should define it.

> re WebID Profile Document Access Privileges:
> 
> The intended access privileges about WebID Profile Documents is clarified in [w3c/WebID#7](https://github.com/w3c/WebID/pull/7) ( as per [w3c/WebID#1](https://github.com/w3c/WebID/issues/1) ) in that the WebID 1.0 specification does not set a constraint - documents can be public or protected, and they may contain public or private information.

Agree, we should not place requirements on the permissions on a WebID/Solid Profile Document.

**Comment by elf-pavlik** [comment]
> it SHOULD have one or more `space:storage` triples

I raised specific a issue about `space:storage` #51 

I see it similar to `solid:oidcIssuer` conversation which led to:

> @elf-pavlik - or perhaps you are right that it does not belong here at all. If it is removed, then the dependency tree would go WebID Spec <- Solid Profile Spec <- OIDC/InterOp/Index/New Specs. So my current thinking is remove solid:oidcIssuer, point to the WebID Spec for all conformance rules except the solid:profileDocument predicate. Explain in non-normative text that Solid-OIDC is the currently most widely used identity protocol and that therefore the solid:oidcIssuer can be expected in the WebID document.

[Solid Application Interoperability](https://solid.github.io/data-interoperability-panel/specification/) does **NOT** use `space:storage` at all. Instead, as I explain in #51 it defines two distinct paths. One for Authorization Agent and the second for other Applications (TL;DR go to Authorization Agent so that the user can authorize your access first, since even the discovery is limited to just what that app access). 

I think that whatever (mini)spec relies on `space:storage` for discovery, and however it takes authorization into account (if at all), it should extend Solid Profile Spec in the same way SAI will be extending it.

If more than one discovery specs rely on that predicate, it might be justified to propose it up the spec chain, I don't think this is the case for `space:storage`. 

-----------
> I believe this Other Predicates should be a separate living document which can be amended as different specifications develop. It would have a list of predicates used in the WebID document and Profile document. This would be a lightweight version of something like IANA Link Relations registry https://www.iana.org/assignments/link-relations/link-relations.xhtml

I still think this would be the cleanest approach for documenting what can be expected in Solid Profiles without adding:

> it *SHOULD* have one or more `space:storage` triples

Just to potentially realize a year or two later that *SHOULD NOT* would have been more appropriate.

**Comment by jeff-zucker** [comment]
> > I believe this Other Predicates should be a separate living document which can be amended as different specifications develop. It would have a list of predicates used in the WebID document and Profile document. This would be a lightweight version of something like IANA Link Relations registry https://www.iana.org/assignments/link-relations/link-relations.xhtml
> 

An interesting idea, we'll discuss.

> > it _SHOULD_ have one or more `space:storage` triples

That statement will not occur in isolation, rather will occur in the context of a description of why a person might or might not want their storage(s) findable.

**Comment by jeff-zucker** [comment]
Closing this issue.  It should have been a discussion rather than an issue.  Many of the points here have been taken up in other issues and there is no particular proposal in this issue which needs to be acted on.  However, useful discussion for historical purposes.

**Comment by timbl** [comment]
If you close this as the points have been taken up in other issues, please link to this issues.

**Comment by jeff-zucker** [comment]
@timbl - My understanding is that @VirginiaBalseiro is working on a new set of conformance rules (see https://github.com/solid/webid-profile/issues/89) and I did not feel it would be helpful to bring up issues until those are available for review.  That way we can start from a review of what she proposes rather than with this issue which became muddled.  If her conformance rules don't address my concerns, I will certainly raise the issues again.
